### PR TITLE
Configure and introduce TypeScript

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR configures and introduces [TypeScript](https://www.typescriptlang.org). As the first file to serve as an example, `src/blocks/homepage-articles/utils.ts` is used. 
What was done here was renaming it (`*.js` -> `*.ts`) and adding type information. 

### How to test the changes in this Pull Request:

1. There should be no functional changes, this is a refactoring
2. Have a look at new TS code – does it make sense?
3. Observe that the new CI step – `typescript-check` is passing

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by adekbadek** [comment]
Updated to use `newspack-scripts` for TS validation. Please have a look @Automattic/newspack-product

File: src/blocks/homepage-articles/utils.ts [hunk 1]
```diff
@@ -38,29 +37,47 @@ const POST_QUERY_ATTRIBUTES = [
...
  * A reflow should happen if:
  * 1. Query-changing attributes of a block change
  * 2. The top-level blocks order changes. A Homepage Articles
  *    block might be nested somewhere.
- *
- * @param {Object} prevProps Edit component props
- * @param {Object} props     Edit component props
  */
-export const shouldReflow = ( prevProps, props ) =>
+export const shouldReflow = ( prevProps: HomepageArticlesProps, props: HomepageArticlesProps ) =>
```
Comment (miguelpeixe) [inline src/blocks/homepage-articles/utils.ts:71]:
```suggestion
export const shouldReflow = ( prevProps: HomepageArticlesProps, props: HomepageArticlesProps ): boolean =>
```

**Review by miguelpeixe — CHANGES_REQUESTED** [review]
:tada: I really like how this is put together!

Left a nit and a question below. Also, wonder how verbose we should be regarding these function return types.

It feels like the entire `utils.ts` could use a better inline doc, and by following that protocol, the return types should also be declared for all functions, even if just a broad `object`. WDYT?

File: src/types/index.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+declare global {
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply src/types/index.ts:1]: Interesting idea – I'd say depending on the size of this, which would inform the tradeoff between reusability and maintenance burden. A type like `Block` in this file should ultimately live in the [Gutenberg repository](https://github.com/WordPress/gutenberg), but I can't find such a definition there. So there's two ways we can go about it, and I think for now this should be considered later, once we get an idea of how large would the shared type declarations be.

[the remark this answers is not in the document]
Comment (dkoo) [reply src/types/index.ts:1]: I agree with the idea of putting it in `newspack-scripts`, as many of these ilke `PostId` will need to be duplicated across repositories. But I'm also okay with deferring that until we start introducing TS to other repos.

**Review by dkoo — APPROVED** [review]
This looks good to me, although I'm not yet fully versed in TypeScript and its best practices. It intuitively makes sense, for the most part. Thanks for the implementation and example conversion of the utils module, which will serve as a valuable template for building and converting more modules going forward.

**Review by miguelpeixe — APPROVED** [review]
There are still a couple of questions I made on my initial review but they are non-blocking:

> It feels like the entire utils.ts could use a better inline doc, and by following that protocol, the return types should also be declared for all functions, even if just a broad object. WDYT?

> On type declaration, should we follow a standard? This proposal contains types declared on a separate file and inline. Maybe a src/blocks/homepage-articles/types.ts could house all of the block-related types?

**Comment by adekbadek** [comment]
Sorry @miguelpeixe, I've forgotten to reply fully 🙈 

> It feels like the entire utils.ts could use a better inline doc

Added some comments in 017603d300122a37f623e2cdd213b051eae97eb0

> the return types should also be declared for all functions, even if just a broad object. 

Not necessarily! [TypeScript will infer](https://www.typescriptlang.org/docs/handbook/type-inference.html) a lot of types, and if it can't (or there's a discrepancy) it will error out on validation. 

> On type declaration, should we follow a standard? This proposal contains types declared on a separate file and inline. 

I don't see using types in a separate file _and_ inline as an issue – some types apply to a single function only, some are reused. Do you think no types should be declared inline?

> Maybe a src/blocks/homepage-articles/types.ts could house all of the block-related types?

`src/types/index.ts` has that role, but it's not in `src/blocks/homepage-articles` because these types are universal for all blocks, not just this one. Does that address this concern?

---

If we're discussing conventions, I was wondering about a naming convention for all types, to make grepping easier. Something like suffixing all types with a `T` (e.g. `HomepageArticlesAttributesT`). Regular variables are in camel case, but React components are pascal-cased, like types. This would help to avoid potential confusion of the two. WDYT @miguelpeixe @dkoo ?

**Review by miguelpeixe — APPROVED** [review]
Thank you for the investigations, @adekbadek!

> Do you think no types should be declared inline?

I don't mind it, it was more of a question to understand standards and how we should continue implementing TS moving forward.

**Comment by dkoo** [comment]
I'd say we should be able to merge this and continue discussion on best practices as we develop new features.

**Comment by adekbadek** [comment]
I agree! I was planning to merge this after last approval, but only now found time to resolve conflicts.

**Pull request merged**